### PR TITLE
Extrusion shaders

### PR DESCRIPTION
**Pull request by lbud**
Companion PR to https://github.com/mapbox/mapbox-gl-js/pull/3223.

Adds
- extrusion shader: used to draw solid colored extruded features 
- extrusion pattern shader: used to draw patterned extruded features
- extrusion texture shader: used to render the entire extrusion framebuffer back to the map at a desired opacity

cc @nickidlugash @ansis @lucaswoj @jfirebaugh

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -11,6 +11,18 @@ module.exports = {
...
   },
+  extrusion: {
+      fragmentSource: fs.readFileSync(path.join(__dirname, 'src/extrusion.fragment.glsl'), 'utf8'),
+      vertexSource: fs.readFileSync(path.join(__dirname, 'src/extrusion.vertex.glsl'), 'utf8')
+  },
+  extrusionpattern: {
+      fragmentSource: fs.readFileSync(path.join(__dirname, 'src/extrusionpattern.fragment.glsl'), 'utf8'),
+      vertexSource: fs.readFileSync(path.join(__dirname, 'src/extrusionpattern.vertex.glsl'), 'utf8')
+  },
+  extrusiontexture: {
```
Comment (lucaswoj) [inline index.js]: It took me a few moments to understand what `extrusiontexture` is. A name that captures the second-pass-ness of this shader would be ideal. I'll sleep on it.

File: src/extrusion.fragment.glsl [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+#ifdef MAPBOX_GL_JS
+#pragma mapbox: define lowp float minH
+#pragma mapbox: define lowp float maxH
+#endif
+#pragma mapbox: define lowp vec4 color
+
+void main() {
+#ifdef MAPBOX_GL_JS
+    #pragma mapbox: initialize lowp float minH
+    #pragma mapbox: initialize lowp float maxH
```
Comment (lucaswoj) [inline src/extrusion.fragment.glsl]: What does the `H` stand for?

Comment (lucaswoj) [inline src/extrusion.fragment.glsl]: Our naming convention for identifiers in shaders is `lower_underscore_case`

File: src/extrusion.vertex.glsl [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+        directional *= clamp((t + minH) * pow(maxH / 150.0, 0.5), mix(0.7, 0.98, 1.0 - u_lightintensity), 1.0);
+    }
+
+    // Assign final color based on surface + ambient light color, diffuse light directional, and light color
+    // with lower bounds adjusted to hue of light
+    // so that shading is tinted with the complementary (opposite) color to the light color
+    v_color.r += clamp(color.r * directional * u_lightcolor.r, mix(0.0, 0.3, 1.0 - u_lightcolor.r), 1.0);
+    v_color.g += clamp(color.g * directional * u_lightcolor.g, mix(0.0, 0.3, 1.0 - u_lightcolor.g), 1.0);
+    v_color.b += clamp(color.b * directional * u_lightcolor.b, mix(0.0, 0.3, 1.0 - u_lightcolor.b), 1.0);
+}
```
Comment (lucaswoj) [inline src/extrusion.vertex.glsl]: I take it from the `// TODO`s and commented-out code that this file is still somewhat a WIP 😄

File: src/extrusionpattern.vertex.glsl [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    float z = t > 0.0 ? maxH : minH;
+
+    gl_Position = u_matrix * vec4(a_pos, z, 1);
+
+    vec2 scaled_size_a = u_scale_a * u_pattern_size_a;
+    vec2 scaled_size_b = u_scale_b * u_pattern_size_b;
+
+    // the following offset calculation is duplicated from the regular pattern shader:
+    vec2 offset_a = mod(mod(mod(u_pixel_coord_upper, scaled_size_a) * 256.0, scaled_size_a) * 256.0 + u_pixel_coord_lower, scaled_size_a);
+    vec2 offset_b = mod(mod(mod(u_pixel_coord_upper, scaled_size_b) * 256.0, scaled_size_b) * 256.0 + u_pixel_coord_lower, scaled_size_b);
```
Comment (lucaswoj) [inline src/extrusionpattern.vertex.glsl]: We can factor this shared out into `util.glsl`

**Review by jfirebaugh — APPROVED** [review]
🎉

**Pull request merged**